### ISSUE TITLE
fmplayer fails to compile with ld giving "multiple definitions" error

### ISSUE THREAD

**Issue by kir12** [myon98/98fmplayer#3]
Hi, so I attempted to compile fmplayer on Arch Linux and ended up getting the below error(s):

```
  CCLD     98fmplayer
/usr/bin/ld: ../fmdsp/98fmplayer-fmdsp-pacc.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/../fmdsp/fmdsp-pacc.h:10: multiple definition of `opna'; 98fmplayer-main.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/./../fmdsp/fmdsp-pacc.h:10: first defined here
/usr/bin/ld: ../fmdsp/98fmplayer-fmdsp-pacc.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/../fmdsp/fmdsp-pacc.h:9: multiple definition of `work'; 98fmplayer-main.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/./../fmdsp/fmdsp-pacc.h:9: first defined here
collect2: error: ld returned 1 exit status

```

This error comes up after running `make` as part of the list of commands you specified inside the readme for compiling fmplayer using GTK:

```
$ cd gtk
$ autoreconf -i
$ ./configure
$ make
$ ./98fmplayer
```

For your convenience, I'll link the entire output of `make` up to and including error here. (Note that the previous commands all worked with no errors)

```
  CC       98fmplayer-main.o
  CC       98fmplayer-toneview.o
  CC       98fmplayer-wavesave.o
  CC       ../tonedata/98fmplayer-tonedata.o
  CC       ../common/98fmplayer-fmplayer_file.o
  CC       ../common/98fmplayer-fmplayer_file_gio.o
  CC       ../common/98fmplayer-fmplayer_work_opna.o
  CC       ../common/98fmplayer-fmplayer_drumrom_unix.o
  CC       ../common/98fmplayer-fmplayer_fontrom_unix.o
  CC       ../fft/98fmplayer-fft.o
  CC       ../libopna/98fmplayer-opnaadpcm.o
  CC       ../libopna/98fmplayer-opnadrum.o
  CC       ../libopna/98fmplayer-opnafm.o
  CC       ../libopna/98fmplayer-opnassg.o
  CC       ../libopna/98fmplayer-opnassg-sinc-c.o
  CC       ../libopna/98fmplayer-opnatimer.o
  CC       ../libopna/98fmplayer-opna.o
  CC       ../fmdriver/98fmplayer-fmdriver_fmp.o
  CC       ../fmdriver/98fmplayer-fmdriver_pmd.o
  CC       ../fmdriver/98fmplayer-fmdriver_common.o
  CC       ../fmdriver/98fmplayer-ppz8.o
  CC       ../fmdsp/98fmplayer-fmdsp-pacc.o
  CC       ../fmdsp/98fmplayer-font_rom.o
  CC       ../fmdsp/98fmplayer-font_fmdsp_small.o
  CC       ../fmdsp/98fmplayer-fmdsp_platform_unix.o
  CC       ../pacc/98fmplayer-pacc-gl.o
  CC       ../soundout/98fmplayer-soundout.o
  CC       ../soundout/98fmplayer-jackout.o
  CC       ../soundout/98fmplayer-pulseout.o
  CC       ../soundout/98fmplayer-alsaout.o
In file included from ../soundout/alsaout.c:9:
/usr/include/asoundlib.h:1:2: warning: #warning This header is deprecated, use <alsa/asoundlib.h> instead. [-Wcpp]
    1 | #warning This header is deprecated, use <alsa/asoundlib.h> instead.
      |  ^~~~~~~
  CC       98fmplayer-configdialog.o
  CC       98fmplayer-oscilloview-gl.o
  CC       ../libopna/libsse_a-opnassg-sinc-sse2.o
  AR       libsse.a
ar: `u' modifier ignored since `D' is the default (see `U')
  CCLD     98fmplayer
/usr/bin/ld: ../fmdsp/98fmplayer-fmdsp-pacc.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/../fmdsp/fmdsp-pacc.h:10: multiple definition of `opna'; 98fmplayer-main.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/./../fmdsp/fmdsp-pacc.h:10: first defined here
/usr/bin/ld: ../fmdsp/98fmplayer-fmdsp-pacc.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/../fmdsp/fmdsp-pacc.h:9: multiple definition of `work'; 98fmplayer-main.o:/home/brianl/Projects/fmplayer-0.1.14/gtk/./../fmdsp/fmdsp-pacc.h:9: first defined here
collect2: error: ld returned 1 exit status
make: *** [Makefile:660: 98fmplayer] Error 1

```

I verified that I have GTK3 installed, so that's probably not it; I don't *think* fmplayer needs GTK2 but feel free to correct me if I'm wrong. Any ideas on how to resolve this issue?

I should also note that I attempted compiling on both the latest commit on master and the latest *release* (which is a few commits behind), with both states of the codebae giving the same error.

**Comment by gzaffin** [comment]
Hi kir12,
I have a compiled version of fmplayer, but, as an exercise, I strarted again from scratch
```
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~$ git clone https://github.com/takamichih/fmplayer.git   
Cloning into 'fmplayer'...   
remote: Enumerating objects: 1503, done.   
remote: Total 1503 (delta 0), reused 0 (delta 0), pack-reused 1503   
Receiving objects: 100% (1503/1503), 1019.63 KiB | 121.00 KiB/s, done.   
Resolving deltas: 100% (977/977), done.   
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~$ cd fmplayer/   
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~/fmplayer$ cd gtk                                                                                                                    
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~/fmplayer/gtk$ ./autogen.sh
configure.ac:6: installing './ar-lib'                                                                                                                                                        
configure.ac:4: installing './compile'                                                                                                                                                       
configure.ac:2: installing './install-sh'                                                                                                                                                    
configure.ac:2: installing './missing'                                                                                                                                                       
Makefile.am: installing './depcomp'                                                                                                                                                          
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~/fmplayer/gtk$ ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for gcc option to accept ISO C99... none needed
checking for ranlib... ranlib
checking for ar... ar
checking the archiver (ar) interface... ar
checking dependency style of gcc... gcc3
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for JACK... no
checking for PULSE... yes
checking for ALSA... yes
checking for GTK3... yes
checking for SNDFILE... yes
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking emmintrin.h usability... yes
checking emmintrin.h presence... yes
checking for emmintrin.h... yes
checking for glGetString in -lGL... yes
checking GL/glcorearb.h usability... yes
checking GL/glcorearb.h presence... yes
checking for GL/glcorearb.h... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: executing depfiles commands
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~/fmplayer/gtk$ make
  CC       98fmplayer-main.o
  CC       98fmplayer-toneview.o
  CC       98fmplayer-wavesave.o
wavesave.c: In function ‘msgbox_err’:
wavesave.c:17:27: warning: format not a string literal and no format arguments [-Wformat-security]
   17 |                           msg);
      |                           ^~~
  CC       ../tonedata/98fmplayer-tonedata.o
  CC       ../common/98fmplayer-fmplayer_file.o
  CC       ../common/98fmplayer-fmplayer_file_gio.o
  CC       ../common/98fmplayer-fmplayer_work_opna.o
  CC       ../common/98fmplayer-fmplayer_drumrom_unix.o
  CC       ../common/98fmplayer-fmplayer_fontrom_unix.o
  CC       ../fft/98fmplayer-fft.o
  CC       ../libopna/98fmplayer-opnaadpcm.o
  CC       ../libopna/98fmplayer-opnadrum.o
  CC       ../libopna/98fmplayer-opnafm.o
  CC       ../libopna/98fmplayer-opnassg.o
  CC       ../libopna/98fmplayer-opnassg-sinc-c.o
  CC       ../libopna/98fmplayer-opnatimer.o
  CC       ../libopna/98fmplayer-opna.o
  CC       ../fmdriver/98fmplayer-fmdriver_fmp.o
  CC       ../fmdriver/98fmplayer-fmdriver_pmd.o
  CC       ../fmdriver/98fmplayer-fmdriver_common.o
  CC       ../fmdriver/98fmplayer-ppz8.o
  CC       ../fmdsp/98fmplayer-fmdsp-pacc.o
  CC       ../fmdsp/98fmplayer-font_rom.o
  CC       ../fmdsp/98fmplayer-font_fmdsp_small.o
  CC       ../fmdsp/98fmplayer-fmdsp_platform_unix.o
  CC       ../pacc/98fmplayer-pacc-gl.o
  CC       ../soundout/98fmplayer-soundout.o
  CC       ../soundout/98fmplayer-pulseout.o
  CC       ../soundout/98fmplayer-alsaout.o
In file included from ../soundout/alsaout.c:9:
/usr/include/asoundlib.h:1:2: warning: #warning This header is deprecated, use <alsa/asoundlib.h> instead. [-Wcpp]
    1 | #warning This header is deprecated, use <alsa/asoundlib.h> instead.
      |  ^~~~~~~
  CC       98fmplayer-configdialog.o
  CC       98fmplayer-oscilloview-gl.o
  CC       ../libopna/libsse_a-opnassg-sinc-sse2.o
  AR       libsse.a
ar: `u' modifier ignored since `D' is the default (see `U')
  CCLD     98fmplayer
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~/fmplayer/gtk$ ./98fmplayer 

(98fmplayer:6184): dbind-WARNING **: 15:52:40.772: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceUnknown: The name org.a11y.Bus was not provided by any .service files
vertex shader error: 
0:1(10): error: GLSL 1.50 is not supported. Supported versions are: 1.10, 1.20, and 1.00 ES

s:
VERT_IN vec4 coord;
VERT_OUT vec2 texcoord;
void main(void) {
  gl_Position = vec4(coord.xy, 0.0, 1.0);
  texcoord = coord.zw;
}
```
So, vanilla codebase compilation completes on my GNU/Lubuntu linux box, but 98fmplayer needs some configuration tweak for running well. I hope I can pinp-pong with You for finding out minimal useful set of changes needed.

**Comment by kir12** [comment]
I noticed that you used the command `./autogen.sh` instead of `autoreconf -i` -- might I ask why?

**Comment by gzaffin** [comment]
Hi kir12,
before calling autogen.sh,
I did checked that autogen.sh indeed is just
```
#! /bin/bash
autoreconf -i
```

**Comment by kir12** [comment]
Alright, so that's that. Looking through your code output, it appears that the stage inside `make` where I'm failing  is the last step before the procedure completes. My first thought was that it might be a sytnax error inside the source, but that doesn't explain why the source code compiled properly on your end. The error also claims that there's a repeated definition and points to the same file twice, which also doesn't make sense. Any other ideas?

One other option is to set up a Docker container and see if it compiles + runs properly there -- the only problem is that there's no easy way to get user interaction to work well since the container usually only works over command line. (and the existing options of getting visual output like x11 over SSH or a VNC server add unnacceptable amounts of lag)

**Comment by gzaffin** [comment]
Hi kir12,
please try to solve `one definiton rule` error with this patch
```
giangiacomo@giangiacomo-HP-Pavilion-dv6000-RR374EA-ABZ:~/fmplayer/gtk$ git diff ../fmdsp/fmdsp-pacc.h
diff --git a/fmdsp/fmdsp-pacc.h b/fmdsp/fmdsp-pacc.h                                                                                                                                         
index 3e10e03..638e9dd 100644                                                                                                                                                                
--- a/fmdsp/fmdsp-pacc.h                                                                                                                                                                     
+++ b/fmdsp/fmdsp-pacc.h
@@ -6,8 +6,8 @@
 struct fmdsp_pacc;
 struct pacc_ctx;
 struct pacc_vtable;
-struct fmdriver_work work;
-struct opna opna;
+struct fmdriver_work /*work*/;
+struct opna /*opna*/;
 struct fmplayer_fft_input_data;
 struct fmdsp_font;
```
I can think of, but I'm not sure, that in my tool-chain double defined `work` and `opna` abstract data structs are silently optimized out because they are not used at all, they only break one definition rule.

Does it work for You?

**Comment by kir12** [comment]
Yeah, it works compiles perfectly now. (and runs) Very strange why this came up. However, there are still a few problems at least for me:
- The filesystem interface glitches and lags very frequently, making it difficult to retrieve my music files.
- The FM/SSG audio playback is rock solid, but I am finding significant differences between percussion playback -- it seems that either fmplayer's RSS samples are different than the ones used in FMPMD2000, they're abnormally quiet, or fmplayer isn't playing any RSS samples at all. (Any three of which is a definite problem)
- The visual interface itself tends to freeze every so often before picking back up again.
Do you experience any of these issues too?

**Comment by gzaffin** [comment]
Did You let 98fmplayer find ym2608_adpcm_rom.bin as per README.md instructions?
See what there is in `common\fmplayer_drumrom_unix.c`.

Feel free to try and search over the internet for `planetemu YM2608 Internal ROM` to find the rom.bin You need.

**Comment by kir12** [comment]
Yeah you were right, that resolves the RSS issues. The filesystem/interface freezing up is still an open issue though -- how smoothly does playback on your end run?

**Comment by myon98** [comment]
Sorry, it was a very simple mistake on my part, should be fixed with a6772b9.
